### DISCUSSION TITLE
[tampermonkey] GM_getValue widening

### DISCUSSION THREAD

**Discussion by j4k0xb** [DefinitelyTyped/DefinitelyTyped#61719] [Issues with a @types package]
@melusc
Do you think it makes sense to widen the return type that's inferred from `defaultValue`? Since the actual saved value can be different.

E.g. returning `string` instead of `""`:
![image](https://user-images.githubusercontent.com/55899582/184508007-702c5b28-299d-4e85-8bc8-c3caa10abeec.png)

Possibly like this:
```ts
type ToPrimitive<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T;
  
// ...
declare function GM_getValue<TValue>(
    name: string,
    defaultValue?: TValue
): ToPrimitive<TValue>;
// similar in GM.getValue
```

[a comment by an automated account is not shown]

**Comment by melusc** [comment]
That seems sensible, but I'm not sure how to solve scenarios like this, where you wouldn't want the return type to be widened.

![image](https://user-images.githubusercontent.com/61409445/184509063-3bd9f4a3-19e4-495f-a82a-020db17b63bc.png)


I copied the types for `GM.getValue` from `GM_getValue`, so maybe the owners can chime in?

**Comment by nikolay-borzov** [comment]
I haven't consider case like this.
```ts
const token = GM_getValue('token', '');
  \/ This condition will always return 'false' since the types '""' and '"abs"' have no overlap.      
if(token === 'abs') { }
```
But as I can see it can be resolved by specifying type explicitly:
```ts
const token = GM_getValue<string>('token', '');
if(token === 'abs') { }
```
I think here it's fine to be explicit
